File: frontend/src/api/analytics.ts
/**
 * Analytics API client — MB.6 analytics endpoints.
 *
 * GET /analytics/geo               — server-side geo points for the Leaflet map (#20).
 *   SECURITY: never call external ip-api.com or similar CDN geo services.
 *   The server resolves coordinates; the UI just renders them on a Leaflet map.
 *   Air-gap-safe by design (ADR-0029 D1, issue #20).
 *
 * GET /analytics/summary           — aggregate stats for the charts panel.
 * GET /analytics/categories-timeline — category-over-time buckets.
 * GET /analytics/asn               — ranked ASN aggregation (issue #533, A2).
 * GET /analytics/asn/{asn}/narration — local-LLM ASN narrative (issue #533, A2 EARS-5).
 */

import { ApiError, buildHeaders, resolveBaseUrl, assertLoopbackBase } from './client'
import type {
  GeoPoint,
  AnalyticsSummary,
  CategoryTimelineBucket,
  AttackDispositionRow,
  AsnRow,
  AsnNarrationResult,
} from './types'

/**
 * Resolve BASE_URL from environment via the shared helper (fix #81).
 * Routes through client.ts resolveBaseUrl so the loopback guard is applied
 * in prod — identical fail-closed posture to client.ts (ADR-0026).
 */
const BASE_URL = resolveBaseUrl(
  (import.meta as { env?: { VITE_API_BASE_URL?: string; DEV?: boolean; PROD?: boolean } }).env ??
    {},
)

// ADR-0026 loopback guard — fail-closed in prod, same as client.ts (#81).
if (
  (import.meta as { env?: { PROD?: boolean } }).env?.PROD === true &&
  BASE_URL !== ''
) {
  assertLoopbackBase(BASE_URL)
}

async function parseError(res: Response): Promise<ApiError> {
  const detail = await res.json().catch(() => res.text().catch(() => null))
  return new ApiError(res.status, detail, `API ${res.status}: ${res.url}`)
}

/**
 * Fetch server-side geo points for the Leaflet map.
 * GET /analytics/geo
 *
 * All coordinates are resolved server-side (#20). The UI MUST NOT make any
 * client-side calls to ip-api.com or any external geo service — air-gap-safe.
 */
export async function fetchGeo(): Promise<GeoPoint[]> {
  const res = await fetch(`${BASE_URL}/analytics/geo`, { method: 'GET', headers: buildHeaders() })
  if (!res.ok) throw await parseError(res)
  return res.json() as Promise<GeoPoint[]>
}

/**
 * Fetch aggregate analytics summary.
 * GET /analytics/summary
 */
export async function fetchAnalyticsSummary(): Promise<AnalyticsSummary> {
  const res = await fetch(`${BASE_URL}/analytics/summary`, {
    method: 'GET',
    headers: buildHeaders(),
  })
  if (!res.ok) throw await parseError(res)
  return res.json() as Promise<AnalyticsSummary>
}

/**
 * Fetch category-over-time timeline.
 * GET /analytics/categories-timeline
 */
export async function fetchCategoriesTimeline(): Promise<CategoryTimelineBucket[]> {
  const res = await fetch(`${BASE_URL}/analytics/categories-timeline`, {
    method: 'GET',
    headers: buildHeaders(),
  })
  if (!res.ok) throw await parseError(res)
  return res.json() as Promise<CategoryTimelineBucket[]>
}

/**
 * Fetch attack-category × disposition cross-tab.
 * GET /analytics/attack-dispositions
 *
 * Returns [{attack_type, action, count}] for top-5 attack categories + Other.
 * Empty array when no categorized events exist (degrade-to-hidden semantics).
 *
 * Issue #214 — additive endpoint (ADR-0029 D1).
 */
export async function fetchAttackDispositions(): Promise<AttackDispositionRow[]> {
  const res = await fetch(`${BASE_URL}/analytics/attack-dispositions`, {
    method: 'GET',
    headers: buildHeaders(),
  })
  if (!res.ok) throw await parseError(res)
  return res.json() as Promise<AttackDispositionRow[]>
}

/**
 * Fetch ranked ASN aggregation for the infrastructure lens.
 * GET /analytics/asn
 *
 * Returns [{asn, as_name, total_events, distinct_ips, blocked, blocked_pct}]
 * ordered by total_events descending.
 *
 * Issue #533 A2 — EARS-2.  Zero-egress (ADR-0022/0047).
 */
export async function fetchAsnStats(topN = 15): Promise<AsnRow[]> {
  const res = await fetch(`${BASE_URL}/analytics/asn?top_n=${topN}`, {
    method: 'GET',
    headers: buildHeaders(),
  })
  if (!res.ok) throw await parseError(res)
  return res.json() as Promise<AsnRow[]>
}

/**
 * Fetch a local-LLM narrative for the given ASN.
 * GET /analytics/asn/{asn}/narration[?ai=false]
 *
 * Reuses the ML-7 narration path (ADR-0043).  Pass includeAi=false to
 * get a deterministic rule-only summary without calling the LLM.
 *
 * Issue #533 A2 — EARS-5.  Zero-egress (ADR-0022/0047).
 */
export async function fetchAsnNarration(
  asn: number,
  includeAi = true,
): Promise<AsnNarrationResult> {
  const url = `${BASE_URL}/analytics/asn/${asn}/narration${includeAi ? '' : '?ai=false'}`
  const res = await fetch(url, { method: 'GET', headers: buildHeaders() })
  if (!res.ok) throw await parseError(res)
  return res.json() as Promise<AsnNarrationResult>
}

File: packages/firewatch-api/src/firewatch_api/routes/analytics.py
"""Analytics routes — GET /analytics/* (ADR-0029 D1, MB.1).

Thin controllers delegating to store methods.  No business logic here.

OCSF field mapping is preserved in the store's output (ADR-0029 D4 / ADR-0020):
the API is the view at the boundary, not a second mapping layer.

Issue #533 (A2): adds /analytics/asn (ranked ASN aggregation) and
/analytics/asn/{asn}/narration (one-click local-LLM narrative for an ASN,
reusing ML-7 narration helpers from firewatch_core.ai.narration).

Imports only firewatch-sdk and firewatch-core. Never imports legacy/.
"""
from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query

from firewatch_api.deps import get_event_store, get_pipeline, parse_iso_or_422

logger = logging.getLogger("firewatch.api.analytics")

router = APIRouter(prefix="/analytics", tags=["analytics"])


def _require_store(store: Any) -> Any:
    """Raise 503 if the event store is not available."""
    if store is None:
        raise HTTPException(status_code=503, detail="Event store not available")
    return store


@router.get(
    "/geo",
    summary="Server-side geo points for mapped IPs",
)
async def get_analytics_geo(
    store: Any = Depends(get_event_store),
) -> list[dict[str, Any]]:
    """Return geo-enriched IP points from the ip_geo cache."""
    _require_store(store)
    return await store.get_analytics_geo()  # type: ignore[return-value]


@router.get(
    "/summary",
    summary="Analytics aggregate (total IPs, events, blocked, top country/rule)",
)
async def get_analytics_summary(
    store: Any = Depends(get_event_store),
) -> dict[str, Any]:
    """Return an analytics aggregate dict."""
    _require_store(store)
    return await store.get_analytics_summary()  # type: ignore[return-value]


@router.get(
    "/categories-timeline",
    summary="Blocked events per category per time period",
)
async def get_categories_timeline(
    start: str | None = Query(default=None, description="ISO datetime or date string"),
    end: str | None = Query(default=None, description="ISO datetime or date string"),
    store: Any = Depends(get_event_store),
) -> list[dict[str, Any]]:
    """Return blocked events per category per period (daily/hourly).

    Raises 422 when ``start`` or ``end`` is not a valid ISO-8601 datetime
    (ADR-0029 D3: query-validation failures must return 422, not 500).
    """
    _require_store(store)
    start = parse_iso_or_422("start", start)
    end = parse_iso_or_422("end", end)
    return await store.get_categories_timeline(start=start, end=end)  # type: ignore[return-value]


@router.get(
    "/attack-dispositions",
    summary="Cross-tab of attack category × disposition (action) — top-5 + Other",
)
async def get_attack_dispositions(
    top_n: int = Query(default=5, ge=1, le=20, description="Top N attack categories to return"),
    store: Any = Depends(get_event_store),
) -> list[dict[str, Any]]:
    """Return [{attack_type, action, count}] rows for the top-N attack categories.

    Bounded to top_n categories plus an "Other" bucket for the tail (issue #214).
    Covers all observed actions (BLOCK, DROP, ALERT, ALLOW, LOG).
    Returns an empty list when no categorized events exist.

    Additive endpoint — no existing response shapes are changed (ADR-0029 D1).
    """
    _require_store(store)
    return await store.get_attack_dispositions(top_n=top_n)  # type: ignore[return-value]


@router.get(
    "/asn",
    summary="Ranked ASN / infrastructure aggregation (issue #533, A2)",
)
async def get_analytics_asn(
    top_n: int = Query(
        default=15,
        ge=1,
        le=100,
        description="Top N ASNs to return, ordered by total_events descending.",
    ),
    store: Any = Depends(get_event_store),
) -> list[dict[str, Any]]:
    """Return [{asn, as_name, total_events, distinct_ips, blocked, blocked_pct}] rows.

    Groups ip_geo JOIN logs by asn/as_name to expose the infrastructure-level
    threat lens.  Bounded to top_n ASNs (default 15).

    Issue #533 A2 — EARS-2.  Additive endpoint — no existing shapes changed.
    Zero-egress: all data comes from the local ip_geo cache (ADR-0022/0047).
    """
    _require_store(store)
    return await store.get_analytics_asn(top_n=top_n)  # type: ignore[return-value]


@router.get(
    "/asn/{asn}/narration",
    summary="One-click local-LLM narrative for an ASN (issue #533, A2 EARS-5)",
)
async def get_asn_narration(
    asn: int,
    ai: bool = Query(
        default=True,
        description=(
            "Set to false to skip the LLM and return a rule-only summary "
            "(ai_status='skipped', provenance='rule'). "
            "ML-7 EARS-4 degrade contract."
        ),
    ),
    store: Any = Depends(get_event_store),
    pipeline: Any = Depends(get_pipeline),
) -> dict[str, Any]:
    """Return a SHORT narrative for the given ASN grounded in its analytics row.

    Reuses the ML-7 narration helpers (firewatch_core.ai.narration) generalised
    for an ASN entity rather than a single IP.  Degrades to a rule-only narrative
    when the LLM is unavailable (same degrade contract as /threats/{ip}/narration).

    **ADR-0035 provenance:**
        - ``"rule"``  — LLM not called (offline / ai=false).
        - ``"ai"``    — LLM authored the narrative.

    **ADR-0015:** advisory only — no SOAR/block actions.
    **ADR-0022/0047:** zero-egress — local LLM only.

    Returns **404** when the ASN has no stored events.
    Returns **503** when the store is unavailable.
    """
    _require_store(store)

    # Fetch the aggregated row for this ASN (scan top-100 to find it).
    rows: list[dict[str, Any]] = await store.get_analytics_asn(top_n=100)
    asn_row: dict[str, Any] | None = next(
        (r for r in rows if r.get("asn") == asn), None
    )
    if asn_row is None:
        raise HTTPException(
            status_code=404,
            detail=f"ASN {asn} has no stored events.",
        )

    from firewatch_core.ai.narration import (  # local import — no circular risk
        build_asn_narration_prompt,
        build_rule_only_asn_narration,
    )

    if not ai:
        # Caller opted out of AI — return rule-only immediately.
        result = build_rule_only_asn_narration(asn_row)
        result["asn"] = asn
        result["ai_status"] = "skipped"
        return result

    # Determine whether the LLM is available.
    ai_engine = getattr(pipeline, "ai_engine", None) if pipeline is not None else None
    ai_available = ai_engine is not None

    if ai_available:
        try:
            narration_prompt = build_asn_narration_prompt(asn_row)
            asn_label = f"AS{asn}"
            total_events = int(asn_row.get("total_events", 0))
            blocked = int(asn_row.get("blocked", 0))
            samples = [{"rule_id": "asn_narration", "category": "asn_narration",
                        "count": 1, "payload": ""}]
            raw: dict[str, Any] = await ai_engine.analyze_concise(
                ip=asn_label,
                total_events=total_events,
                blocked_events=blocked,
                rules_triggered=1,
                first_seen="",
                last_seen="",
                samples=samples,
                security_mode=False,
                _narration_prompt=narration_prompt,
            )
            narrative_text: str = (
                raw.get("narrative")
                or raw.get("intent")
                or raw.get("executive_summary")
                or ""
            )
            collected: list[str] = []
            if "PROVENANCE:" in narrative_text:
                parts_split = narrative_text.split("PROVENANCE:", 1)
                narrative_text = parts_split[0].strip()
                prov_line = parts_split[1].strip() if len(parts_split) > 1 else ""
                collected = [f.strip() for f in prov_line.split(",") if f.strip()]
            return {
                "asn": asn,
                "narrative": narrative_text,
                "provenance": "ai",
                "collected_fields": collected,
                "ai_status": "ok",
            }
        except Exception:
            logger.warning(
                "analytics.get_asn_narration: LLM call failed for ASN %s — degrading to rule-only",
                asn,
            )

    # Rule-only fallback (AI unavailable or LLM call failed).
    result = build_rule_only_asn_narration(asn_row)
    result["asn"] = asn
    return result
